File: java/src/test/java/scenario/GatewayContext.java
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package scenario;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import io.grpc.ManagedChannel;
import org.hyperledger.fabric.client.ChaincodeEvent;
import org.hyperledger.fabric.client.Checkpointer;
import org.hyperledger.fabric.client.CloseableIterator;
import org.hyperledger.fabric.client.Contract;
import org.hyperledger.fabric.client.Gateway;
import org.hyperledger.fabric.client.InMemoryCheckpointer;
import org.hyperledger.fabric.client.Network;
import org.hyperledger.fabric.client.identity.Identity;
import org.hyperledger.fabric.client.identity.Signer;
import org.hyperledger.fabric.protos.common.Block;
import org.hyperledger.fabric.protos.peer.BlockAndPrivateData;
import org.hyperledger.fabric.protos.peer.FilteredBlock;

public class GatewayContext {
    private final Gateway.Builder gatewayBuilder;
    private final Map<String, EventListener<ChaincodeEvent>> chaincodeEventListeners = new HashMap<>();
    private final Map<String, EventListener<Block>> blockEventListeners = new HashMap<>();
    private final Map<String, EventListener<FilteredBlock>> filteredBlockEventListeners = new HashMap<>();
    private final Map<String, EventListener<BlockAndPrivateData>> blockAndPrivateDataEventListeners = new HashMap<>();
    private Checkpointer checkpointer;
    private ManagedChannel channel;
    private Gateway gateway;
    private Network network;
    private Contract contract;

    public GatewayContext(Identity identity) {
        gatewayBuilder = Gateway.newInstance()
                .identity(identity);
    }

    public GatewayContext(Identity identity, Signer signer) {
        gatewayBuilder = Gateway.newInstance()
                .identity(identity)
                .signer(signer);
    }

    public void connect(ManagedChannel channel) {
        this.channel = channel;
        gateway = gatewayBuilder.connection(channel).connect();
    }

    public void useNetwork(String networkName) {
        network = gateway.getNetwork(networkName);
    }

    public void useContract(String contractName) {
        contract = network.getContract(contractName);
    }

    public void createCheckpointer() {
        checkpointer = new InMemoryCheckpointer();
    }

    public TransactionInvocation newTransaction(String action, String transactionName) {
        if (action.equals("submit")) {
            return TransactionInvocation.prepareToSubmit(gateway, contract, transactionName);
        } else {
            return TransactionInvocation.prepareToEvaluate(gateway, contract, transactionName);
        }
    }

    public void listenForChaincodeEvents(String listenerName, String chaincodeName) {
        receiveChaincodeEvents(listenerName, network.getChaincodeEvents(chaincodeName));
    }

    public void listenForChaincodeEventsUsingCheckpointer(String listenerName, String chaincodeName) {
        CloseableIterator<ChaincodeEvent> iter = network.newChaincodeEventsRequest(chaincodeName)
                .checkpoint(checkpointer)
                .build()
                .getEvents();
        receiveChaincodeEventsUsingCheckpointer(listenerName, iter);
    }

    public void replayChaincodeEvents(String listenerName, String chaincodeName, long startBlock) {
        CloseableIterator<ChaincodeEvent> iter = network.newChaincodeEventsRequest(chaincodeName)
                .startBlock(startBlock)
                .build()
                .getEvents();
        receiveChaincodeEvents(listenerName, iter);
    }

    private void receiveChaincodeEventsUsingCheckpointer(final String listenerName, final CloseableIterator<ChaincodeEvent> iter) {
        closeChaincodeEvents(listenerName);
        EventListener<ChaincodeEvent> e = new CheckpointEventListener<>(iter, checkpointer::checkpointChaincodeEvent);
        chaincodeEventListeners.put(listenerName, e);
    }

    private void receiveChaincodeEvents(final String listenerName, final CloseableIterator<ChaincodeEvent> iter) {
        closeChaincodeEvents(listenerName);
        chaincodeEventListeners.put(listenerName, new BasicEventListener<>(iter));
    }

    public ChaincodeEvent nextChaincodeEvent(String listenerName) throws InterruptedException, IOException {
        return chaincodeEventListeners.get(listenerName).next();
    }

    public void listenForBlockEvents(String listenerName) {
        receiveBlockEvents(listenerName, network.getBlockEvents());
    }

    public void listenForBlockEventsUsingCheckpointer(String listenerName) {
        CloseableIterator<Block> iter = network.newBlockEventsRequest()
                .checkpoint(checkpointer)
                .build()
                .getEvents();
        receiveBlockEventsUsingCheckpointer(listenerName, iter);
    }

    private void receiveBlockEventsUsingCheckpointer(final String listenerName, final CloseableIterator<Block> iter) {
        closeBlockEvents(listenerName);
        EventListener<Block> e = new CheckpointEventListener<>(iter, event-> checkpointer.checkpointBlock(event.getHeader().getNumber()));
        blockEventListeners.put(listenerName, e);
    }

    public void listenForFilteredBlockEventsUsingCheckpointer(String listenerName) {
        CloseableIterator<FilteredBlock> iter = network.newFilteredBlockEventsRequest()
                .checkpoint(checkpointer)
                .build()
                .getEvents();
        receiveFilteredBlockEventsUsingCheckpointer(listenerName, iter);
    }

    private void receiveFilteredBlockEventsUsingCheckpointer(final String listenerName, final CloseableIterator<FilteredBlock> iter) {
        closeFilteredBlockEvents(listenerName);
        EventListener<FilteredBlock> e = new CheckpointEventListener<>(iter, event-> checkpointer.checkpointBlock(event.getNumber()));
        filteredBlockEventListeners.put(listenerName, e);
    }

    public void  listenForBlockAndPrivateDataUsingCheckpointer(String listenerName) {
        CloseableIterator<BlockAndPrivateData> iter = network.newBlockAndPrivateDataEventsRequest()
                .checkpoint(checkpointer)
                .build()
                .getEvents();
        receiveBlockAndPrivateDataEventsUsingCheckpointer(listenerName, iter);
    }

    private void receiveBlockAndPrivateDataEventsUsingCheckpointer(final String listenerName, final CloseableIterator<BlockAndPrivateData> iter) {
        closeBlockAndPrivateDataEvents(listenerName);
        EventListener<BlockAndPrivateData> e = new CheckpointEventListener<>(iter, event-> checkpointer.checkpointBlock(event.getBlock().getHeader().getNumber()));
        blockAndPrivateDataEventListeners.put(listenerName, e);
    }

    public void replayBlockEvents(String listenerName, long startBlock) {
        CloseableIterator<Block> iter = network.newBlockEventsRequest()
                .startBlock(startBlock)
                .build()
                .getEvents();
        receiveBlockEvents(listenerName, iter);
    }

    private void receiveBlockEvents(final String listenerName, final CloseableIterator<Block> iter) {
        closeBlockEvents(listenerName);
        blockEventListeners.put(listenerName, new BasicEventListener<>(iter));
    }

    public Block nextBlockEvent(String listenerName) throws InterruptedException, IOException {
        return blockEventListeners.get(listenerName).next();
    }

    public void listenForFilteredBlockEvents(String listenerName) {
        receiveFilteredBlockEvents(listenerName, network.getFilteredBlockEvents());
    }

    public void replayFilteredBlockEvents(String listenerName, long startBlock) {
        CloseableIterator<FilteredBlock> iter = network.newFilteredBlockEventsRequest()
                .startBlock(startBlock)
                .build()
                .getEvents();
        receiveFilteredBlockEvents(listenerName, iter);
    }

    private void receiveFilteredBlockEvents(final String listenerName, final CloseableIterator<FilteredBlock> iter) {
        closeFilteredBlockEvents(listenerName);
        filteredBlockEventListeners.put(listenerName, new BasicEventListener<>(iter));
    }

    public FilteredBlock nextFilteredBlockEvent(String listenerName) throws InterruptedException, IOException {
        return filteredBlockEventListeners.get(listenerName).next();
    }

    public void listenForBlockAndPrivateDataEvents(String listenerName) {
        receiveBlockAndPrivateDataEvents(listenerName, network.getBlockAndPrivateDataEvents());
    }

    public void replayBlockAndPrivateDataEvents(String listenerName, long startBlock) {
        CloseableIterator<BlockAndPrivateData> iter = network.newBlockAndPrivateDataEventsRequest()
                .startBlock(startBlock)
                .build()
                .getEvents();
        receiveBlockAndPrivateDataEvents(listenerName, iter);
    }

    private void receiveBlockAndPrivateDataEvents(final String listenerName, final CloseableIterator<BlockAndPrivateData> iter) {
        closeBlockAndPrivateDataEvents(listenerName);
        blockAndPrivateDataEventListeners.put(listenerName, new BasicEventListener<>(iter));
    }

    public BlockAndPrivateData nextBlockAndPrivateDataEvent(String listenerName) throws InterruptedException, IOException {
        return blockAndPrivateDataEventListeners.get(listenerName).next();
    }

    public void close() {
        chaincodeEventListeners.values().forEach(EventListener::close);
        blockEventListeners.values().forEach(EventListener::close);
        filteredBlockEventListeners.values().forEach(EventListener::close);
        blockAndPrivateDataEventListeners.values().forEach(EventListener::close);

        if (gateway != null) {
            gateway.close();
        }

        closeChannel();
    }

    public void closeChaincodeEvents(String listenerName) {
        closeEventListener(chaincodeEventListeners, listenerName);
    }

    public void closeBlockEvents(String listenerName) {
        closeEventListener(blockEventListeners, listenerName);
    }

    public void closeFilteredBlockEvents(String listenerName) {
        closeEventListener(filteredBlockEventListeners, listenerName);
    }

    public void closeBlockAndPrivateDataEvents(String listenerName) {
        closeEventListener(blockAndPrivateDataEventListeners, listenerName);
    }

    private <T> void closeEventListener(final Map<String, EventListener<T>> listeners, final String listenerName) {
        listeners.computeIfPresent(listenerName, (name, listener) -> {
            listener.close();
            return null;
        });
    }

    private void closeChannel() {
        if (channel.isShutdown()) {
            return;
        }

        try {
            channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: scenario/node/src/gatewaycontext.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as grpc from '@grpc/grpc-js';
import { BlockEventsOptions, ChaincodeEvent, ChaincodeEventsOptions, Checkpointer, checkpointers, connect, ConnectOptions, Contract, Gateway, Identity, Network, Signer } from '@hyperledger/fabric-gateway';
import { CheckpointEventListener } from './checkpointeventlistener';
import { BaseEventListener } from './baseeventlistener';
import { TransactionInvocation } from './transactioninvocation';
import { assertDefined } from './utils';
import { common, peer } from '@hyperledger/fabric-protos';
import { EventListener } from './eventlistener';

export class GatewayContext {
    readonly #identity: Identity;
    readonly #signer?: Signer;
    readonly #signerClose?: () => void;
    #client?: grpc.Client;
    #gateway?: Gateway;
    #network?: Network;
    #contract?: Contract;
    #checkpointer?: Checkpointer;
    readonly #chaincodeEventListeners: Map<string, EventListener<ChaincodeEvent>> = new Map();
    readonly #blockEventListeners: Map<string, EventListener<common.Block>> = new Map();
    readonly #filteredBlockEventListeners: Map<string, EventListener<peer.FilteredBlock>> = new Map();
    readonly #blockAndPrivateDataEventListeners: Map<string, EventListener<peer.BlockAndPrivateData>> = new Map();

    constructor(identity: Identity, signer?: Signer, signerClose?: () => void) {
        this.#identity = identity;
        this.#signer = signer;
        this.#signerClose = signerClose;
    }

    connect(client: grpc.Client): void {
        this.#client = client;
        const options: ConnectOptions = {
            signer: this.#signer,
            identity: this.#identity,
            client,
        };
        this.#gateway = connect(options);
    }

    useNetwork(channelName: string): void {
        this.#network = this.getGateway().getNetwork(channelName);
        this.#contract = undefined;
    }

    useContract(contractName: string): void {
        this.#contract = this.getNetwork().getContract(contractName);
    }

    newTransaction(action: string, transactionName: string): TransactionInvocation {
        return new TransactionInvocation(action, this.getGateway(), this.getContract(), transactionName);
    }

    createCheckpointer(): void {
        this.#checkpointer = checkpointers.inMemory();
    }

    getCheckpointer(): Checkpointer {
        return assertDefined(this.#checkpointer, 'checkpointer');
    }

    async listenForChaincodeEvents(listenerName: string, chaincodeName: string, options?: ChaincodeEventsOptions): Promise<void> {
        this.closeChaincodeEvents(listenerName);
        const events = await this.getNetwork().getChaincodeEvents(chaincodeName, options);
        const listener = new BaseEventListener(events);
        this.#chaincodeEventListeners.set(listenerName, listener);
    }

    async listenForChaincodeEventsUsingCheckpointer(listenerName: string, chaincodeName: string, options?: ChaincodeEventsOptions): Promise<void> {
        this.closeChaincodeEvents(listenerName);
        const events = await this.getNetwork().getChaincodeEvents(chaincodeName, options);
        const listener = new CheckpointEventListener<ChaincodeEvent>(events, async (event: ChaincodeEvent): Promise <void> => {
            await this.getCheckpointer().checkpointChaincodeEvent(event);});
        this.#chaincodeEventListeners.set(listenerName, listener);
    }

    async nextChaincodeEvent(listenerName: string): Promise<ChaincodeEvent> {
        const event: ChaincodeEvent = await this.getChaincodeEventListener(listenerName).next();
        return event;
    }

    async listenForBlockEvents(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeBlockEvents(listenerName);
        const events = await this.getNetwork().getBlockEvents(options);
        const listener = new BaseEventListener<common.Block>(events);
        this.#blockEventListeners.set(listenerName, listener);
    }

    async listenForBlockEventsUsingCheckpointer(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeBlockEvents(listenerName);
        const events = await this.getNetwork().getBlockEvents(options);
        const listener = new CheckpointEventListener<common.Block>(events, async (event: common.Block): Promise<void> => {
            const header = assertDefined(event.getHeader(), 'block header');
            const blockNumber = header.getNumber();
            await this.getCheckpointer().checkpointBlock(BigInt(blockNumber));
        });
        this.#blockEventListeners.set(listenerName, listener);
    }

    async nextBlockEvent(listenerName: string): Promise<unknown> {
        return await this.getBlockEventListener(listenerName).next();
    }

    async listenForFilteredBlockEvents(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeFilteredBlockEvents(listenerName);
        const events = await this.getNetwork().getFilteredBlockEvents(options);
        const listener = new BaseEventListener<peer.FilteredBlock>(events);
        this.#filteredBlockEventListeners.set(listenerName, listener);
    }

    async listenForFilteredBlockEventsUsingCheckpointer(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeFilteredBlockEvents(listenerName);
        const events = await this.getNetwork().getFilteredBlockEvents(options);
        const listener = new CheckpointEventListener<peer.FilteredBlock>(events, async (event: peer.FilteredBlock): Promise<void> => {
            const blockNumber = event.getNumber();
            await this.getCheckpointer().checkpointBlock(BigInt(blockNumber));
        });
        this.#filteredBlockEventListeners.set(listenerName, listener);
    }

    async nextFilteredBlockEvent(listenerName: string): Promise<unknown> {
        const event = await this.getFilteredBlockEventListener(listenerName).next();
        return event;
    }
    async listenForBlockAndPrivateDataEvents(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeBlockAndPrivateDataEvents(listenerName);
        const events = await this.getNetwork().getBlockAndPrivateDataEvents(options);
        const listener = new BaseEventListener<peer.BlockAndPrivateData>(events);
        this.#blockAndPrivateDataEventListeners.set(listenerName, listener);
    }

    async listenForBlockAndPrivateDataEventsUsingCheckpointer(listenerName: string, options?: BlockEventsOptions): Promise<void> {
        this.closeBlockAndPrivateDataEvents(listenerName);
        const events = await this.getNetwork().getBlockAndPrivateDataEvents(options);
        const listener = new CheckpointEventListener<peer.BlockAndPrivateData>(events, async (event: peer.BlockAndPrivateData): Promise<void>  => {
            const block = assertDefined(event.getBlock(), 'block');
            const header = assertDefined(block.getHeader(), 'block header');
            const blockNumber = header.getNumber();
            await this.getCheckpointer().checkpointBlock(BigInt(blockNumber));
        });
        this.#blockAndPrivateDataEventListeners.set(listenerName, listener);
    }

    async nextBlockAndPrivateDataEvent(listenerName: string): Promise<unknown> {
        return await this.getBlockAndPrivateDataEventListener(listenerName).next();
    }

    close(): void {
        this.#chaincodeEventListeners.forEach(listener => listener.close());
        this.#blockEventListeners.forEach(listener => listener.close());
        this.#filteredBlockEventListeners.forEach(listener => listener.close());
        this.#blockAndPrivateDataEventListeners.forEach(listener => listener.close());
        this.#gateway?.close();
        this.#client?.close();
        if (this.#signerClose) {
            this.#signerClose();
        }
    }

    closeChaincodeEvents(listenerName: string): void {
        this.#chaincodeEventListeners.get(listenerName)?.close();
        this.#chaincodeEventListeners.delete(listenerName);
    }

    closeBlockEvents(listenerName: string): void {
        this.#blockEventListeners.get(listenerName)?.close();
        this.#blockEventListeners.delete(listenerName);
    }

    closeFilteredBlockEvents(listenerName: string): void {
        this.#filteredBlockEventListeners.get(listenerName)?.close();
        this.#filteredBlockEventListeners.delete(listenerName);
    }

    closeBlockAndPrivateDataEvents(listenerName: string): void {
        this.#blockAndPrivateDataEventListeners.get(listenerName)?.close();
        this.#blockAndPrivateDataEventListeners.delete(listenerName);
    }
    private getGateway(): Gateway {
        return assertDefined(this.#gateway, 'gateway');
    }

    private getNetwork(): Network {
        return assertDefined(this.#network, 'network');
    }

    private getContract(): Contract {
        return assertDefined(this.#contract, 'contract');
    }

    private getChaincodeEventListener(listenerName: string): EventListener<ChaincodeEvent> {
        return assertDefined(this.#chaincodeEventListeners.get(listenerName), `chaincodeEventListener: ${listenerName}`);
    }

    private getBlockEventListener(listenerName: string): EventListener<unknown> {
        return assertDefined(this.#blockEventListeners.get(listenerName), `blockEventListener: ${listenerName}`);
    }

    private getFilteredBlockEventListener(listenerName: string): EventListener<unknown> {
        return assertDefined(this.#filteredBlockEventListeners.get(listenerName), `filteredBlockEventListener: ${listenerName}`);
    }

    private getBlockAndPrivateDataEventListener(listenerName: string): EventListener<unknown> {
        return assertDefined(this.#blockAndPrivateDataEventListeners.get(listenerName), `blockAndPrivateDataEventListener: ${listenerName}`);
    }
}
